File: leetcode/0046.go
package leetcode

func permute(nums []int) (ans[][]int) {
    n := len(nums)
    var dfs func(int)
    dfs = func(i int){
        if i == n {
            ans = append(ans,append([]int{},nums...))
            return
        }
        for j := i; j < n; j++ {
            nums[i],nums[j] = nums[j],nums[i]
            dfs(i+1)
            nums[i],nums[j] = nums[j],nums[i]
        }
    }
    dfs(0)
    return
}

File: leetcode/0046.cpp
class Solution {
public:
    vector<vector<int>> permute(vector<int>& nums) {
        vector<vector<int>> ans;
        int n = nums.size();
        auto dfs = [&](auto&& dfs, auto i) {
            if (i == n) {
                ans.emplace_back(nums);
                return;
            }
            for (auto j = i; j < n; j++) {
                swap(nums[i], nums[j]);
                dfs(dfs, i + 1);
                swap(nums[i], nums[j]);
            }
        };
        dfs(dfs, 0);
        return ans;
    }
};